### PR TITLE
[yul] Add support for EVM version-dependent rules.

### LINKED ISSUE #7542: [Yul] EVMVersion-dependent rules
For selfbalance, it would make sense to add a rule `balance(address()) -> selfbalance()`, but that can only be done if we are running against a post-Istanbul VM. So for implementing this, we need the VM version available when we check for a match - so maybe this can be done inside the precondition function for the rule.

### PR DESCRIPTION
**Pull request by aarlt**
Fixes #7542, #8882

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `99768c2833aa02bb2d2e12b3f24e69b0b479c33d`:
```
Coding style error:
 libyul/optimiser/SimplificationRules.h:145: void addEvmRules(langutil::EVMVersion const &_version);
 libyul/optimiser/SimplificationRules.h:146: void addEvmRules(langutil::EVMVersion const &_version, Rules const& _rules);
 libyul/optimiser/SimplificationRules.h:147: void addEvmRule(langutil::EVMVersion const &_version, Rule const& _rule);

```
Please check that your changes are working as intended.

File: libyul/optimiser/SimplificationRules.h [hunk 1]
```diff
@@ -42,37 +44,6 @@ class Pattern;
 /**
  * Container for all simplification rules.
  */
-class SimplificationRules: public boost::noncopyable
```
Comment (chriseth) [inline libyul/optimiser/SimplificationRules.h:45]: Why did you move the class?

File: libyul/optimiser/SimplificationRules.h [hunk 2]
```diff
@@ -134,5 +105,62 @@ class Pattern
 	unsigned m_matchGroup = 0;
 	std::map<unsigned, Expression const*>* m_matchGroups = nullptr;
 };
+class SimplificationRules: public boost::noncopyable
+{
+public:
+	using Rule = evmasm::SimplificationRule<Pattern>;
+	using Rules = std::vector<Rule>;
```
Comment (chriseth) [inline libyul/optimiser/SimplificationRules.h]: I don't think this and the next make the code easier to read.

File: libyul/optimiser/SimplificationRules.cpp [hunk 3]
```diff
@@ -36,22 +36,44 @@ using namespace solidity::evmasm;
...
-	for (auto const& rule: rules.m_rules[uint8_t(instruction->first)])
+	result = rules.findFirstMatch(_expr, _dialect, _ssaValues, rules.m_rules[uint8_t(instruction->first)]);
+
+	if (result == nullptr)
+		if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&_dialect))
+		{
+			EVMVersion version = evmDialect->evmVersion();
+			rules.addEvmRules(version);
+
+			result = rules.findFirstMatch(_expr, _dialect, _ssaValues, rules.m_evm_rules[version][uint8_t(instruction->first)]);
```
Comment (chriseth) [inline libyul/optimiser/SimplificationRules.cpp]: Why do you need a new member variable for the specialized rules?

File: libyul/optimiser/SimplificationRules.h (hunk 1 above)
Comment (aarlt) [reply libyul/optimiser/SimplificationRules.h:45]: Mainly because of the current usage of the `Pattern` class. Because of the use of forward declaration `Pattern` class was of incomplete type. However, I could change it to a pointer.

File: libyul/optimiser/SimplificationRules.cpp (hunk 3 above)
Comment (aarlt) [reply libyul/optimiser/SimplificationRules.cpp]: I didn't wanted to put the "same" `for` loop twice. But probably it's ok to have these loops.

Comment (chriseth) [reply libyul/optimiser/SimplificationRules.cpp]: I meant to ask why don't you make all rules specialized on an evm version?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `17cd4824a26d58c8f01ece49559828067291f747`:
```
Coding style error:
 libyul/optimiser/SimplificationRules.cpp:93: m_C(std::make_unique<Pattern>(PatternKind::Constant)), m_W(std::make_unique<Pattern>()),
 libyul/optimiser/SimplificationRules.cpp:94: m_X(std::make_unique<Pattern>()), m_Y(std::make_unique<Pattern>()), m_Z(std::make_unique<Pattern>())

```
Please check that your changes are working as intended.

File: libyul/optimiser/SimplificationRules.cpp [hunk 4]
```diff
@@ -76,41 +77,42 @@ std::optional<std::pair<evmasm::Instruction, vector<Expression> const*>>
...
-void SimplificationRules::addRule(SimplificationRule<Pattern> const& _rule)
+void SimplificationRules::addRule(Rule const& _rule)
 {
 	m_rules[uint8_t(_rule.pattern.instruction())].push_back(_rule);
 }
 
-SimplificationRules::SimplificationRules()
+SimplificationRules::SimplificationRules(Dialect const& _dialect)
+	: m_A(std::make_unique<Pattern>(PatternKind::Constant)), m_B(std::make_unique<Pattern>(PatternKind::Constant)),
+	  m_C(std::make_unique<Pattern>(PatternKind::Constant)), m_W(std::make_unique<Pattern>()),
```
Comment (stackenbotten) [inline libyul/optimiser/SimplificationRules.cpp]: Coding style error

File: libyul/optimiser/SimplificationRules.cpp [hunk 5]
```diff
@@ -76,41 +77,42 @@ std::optional<std::pair<evmasm::Instruction, vector<Expression> const*>>
...
+	  m_X(std::make_unique<Pattern>()), m_Y(std::make_unique<Pattern>()), m_Z(std::make_unique<Pattern>())
```
Comment (stackenbotten) [inline libyul/optimiser/SimplificationRules.cpp]: Coding style error

**Review by chriseth — CHANGES_REQUESTED** [review]
This is still a little too complicated for my taste and only works for libyul and not the classic optimizer.

Can you change the functions in RuleList.h to accept an EVMVersion parameter and then - in a regular `simplificationRuleListPartN` function add

```
if (_evmVersion.hasSelfbalance())
  rules.push_back(...);
```
Please not that it should  be `hasSelfbalance` and not a comparison for istanbul!

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `fece50cdac2ca304c5016d8aeea91ec0eb4d4edf`:
```
Coding style error:
 libevmasm/RuleList.h:708: [=]() -> Pattern { return Instruction::SELFBALANCE; },
 libevmasm/RuleList.h:709: false});

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `e84794034c9519979f7225d122270d719e837da3`:
```
Coding style error:
 libevmasm/RuleList.h:722: [=]() -> Pattern { return Instruction::SELFBALANCE; },
 libevmasm/RuleList.h:723: false});

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `ab36603a52964288b60859673fb386cc32463fd7`:
```
Coding style error:
 libevmasm/RuleList.h:723: [=]() -> Pattern { return Instruction::SELFBALANCE; },
 libevmasm/RuleList.h:724: false});

```
Please check that your changes are working as intended.

File: libevmasm/RuleList.h [hunk 6]
```diff
@@ -694,4 +698,32 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
 	return rules;
 }
 
+/// @returns a list of generic simplification rules with additional rules that may apply for the given dialect.
+template <class Pattern>
+std::vector<SimplificationRule<Pattern>> simplificationRuleList(
+	yul::Dialect const& _dialect,
```
Comment (chriseth) [inline libevmasm/RuleList.h]: Ah! Even better!

File: libyul/optimiser/SimplificationRules.cpp [hunk 7]
```diff
@@ -46,7 +46,7 @@ SimplificationRule<yul::Pattern> const* SimplificationRules::findFirstMatch(
 	if (!instruction)
 		return nullptr;
 
-	static SimplificationRules rules;
+	static SimplificationRules rules(_dialect);
```
Comment (chriseth) [inline libyul/optimiser/SimplificationRules.cpp]: This way, the rules list will always use the dialect it was first initialized with. You need a static (maybe LazyInit) variable of type `std::map<EVMVersion, SimplificationRules>`.

File: libevmasm/RuleList.h [hunk 8]
```diff
@@ -694,4 +698,32 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
...
+	Pattern A,
+	Pattern B,
+	Pattern C,
+	Pattern W,
+	Pattern X,
+	Pattern Y,
+	Pattern Z
+)
+{
+	std::vector<SimplificationRule<Pattern>> rules{simplificationRuleList(A, B, C, W, X, Y, Z)};
```
Comment (chriseth) [inline libevmasm/RuleList.h]: Why don't you change that function instead?

File: libyul/optimiser/SimplificationRules.cpp (hunk 7 above)
Comment (aarlt) [reply libyul/optimiser/SimplificationRules.cpp]: Ah ok.. I thought that the compiler will only be initialized with exactly one `EVMVersion`..

File: libevmasm/RuleList.h (hunk 8 above)
Comment (aarlt) [reply libevmasm/RuleList.h]: I thought it may make sense to separate evm version specific rules from generic rules

File: libyul/optimiser/SimplificationRules.cpp (hunk 7 above)
Comment (chriseth) [reply libyul/optimiser/SimplificationRules.cpp]: Never be sure about that :)

File: libevmasm/RuleList.h (hunk 8 above)
Comment (chriseth) [reply libevmasm/RuleList.h]: In a way, we already have a lot of them mixed in, but they do not matter because they do not introduce a new opcode.

Comment (aarlt) [reply libevmasm/RuleList.h]: Ah oki, I will change the function.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `6753cb5eabd6825f07b63d358565d071f8a4008c`:
```
Coding style error:
 libyul/optimiser/SimplificationRules.cpp:58: SimplificationRules &rules = *evmRules[version];

```
Please check that your changes are working as intended.

File: libyul/optimiser/SimplificationRules.cpp [hunk 9]
```diff
@@ -46,7 +46,17 @@ SimplificationRule<yul::Pattern> const* SimplificationRules::findFirstMatch(
...
+	std::optional<EVMVersion> version;
+	if (yul::EVMDialect const* evmDialect = dynamic_cast<yul::EVMDialect const*>(&_dialect))
+		version = evmDialect->evmVersion();
+
+	static std::map<std::optional<EVMVersion>, std::unique_ptr<SimplificationRules>> evmRules;
+
+	if (evmRules.find(version) == evmRules.end())
+		evmRules[version] = std::make_unique<SimplificationRules>(version);
+
+	SimplificationRules &rules = *evmRules[version];
```
Comment (stackenbotten) [inline libyul/optimiser/SimplificationRules.cpp]: Coding style error

**Comment by chriseth** [comment]
This is still missing the changes in the old optimizer.

**Comment by chriseth** [comment]
I'm working on this.

File: libyul/optimiser/SimplificationRules.cpp [hunk 10]
```diff
@@ -46,7 +46,16 @@ SimplificationRule<yul::Pattern> const* SimplificationRules::findFirstMatch(
...
 		return nullptr;
 
-	static SimplificationRules rules;
+	static std::map<std::optional<EVMVersion>, std::unique_ptr<SimplificationRules>> evmRules;
+
+	std::optional<EVMVersion> version;
+	if (yul::EVMDialect const* evmDialect = dynamic_cast<yul::EVMDialect const*>(&_dialect))
+		version = evmDialect->evmVersion();
+
+	if (evmRules.find(version) == evmRules.end())
```
Comment (leonardoalt) [inline libyul/optimiser/SimplificationRules.cpp]:
```suggestion
	if (!evmRules.count(version))
```

**Review by leonardoalt — COMMENTED** [review]
It'd be fun to add a test like
```
{
    let a := address()
    let ret := balance(a)
}
// ====
// EVMVersion: >=istanbul
// ----
// step: expressionSimplifier
//
// { let ret := selfbalance() }
```

Does the rule list work before or after the Yul optimizers in that case?

**Comment by chriseth** [comment]
Oh sorry, forgot to comment here: It turned out that this is not really worth it for the old optimizer, because the rules cannot be used with the old optimizer. The old optimizer can only have rules with instructions that are deterministic. Since `balance` depends on previous calls, it is excluded.

**Comment by chriseth** [comment]
Added the fun test.

**Review by leonardoalt — APPROVED** [review]
